Open to Pull Request form directly on compare page

silverwind opened the pull request:
When creating a PR on Gitea, there is this useless step where one has to click a button on the compare page. Removed it. Note, I'm not yet sure what the relation to `.Flash` here is.

<img width="1339" alt="Screenshot 2024-03-31 at 19 19 39" src="https://github.com/go-gitea/gitea/assets/115237/7d15a8fc-6acc-4da6-ae3a-6aba85a3cc95">

Instead, remove this button and show the form directly. This matches GitHub's behaviour as well.

<img width="1344" alt="Screenshot 2024-03-31 at 19 19 50" src="https://github.com/go-gitea/gitea/assets/115237/d490d134-a455-440c-8f37-14e5e582771f">

delvh commented on templates/repo/diff/compare.tmpl:
```diff
@@ -171,10 +171,7 @@
 	{{if .IsNothingToCompare}}
 		{{if and $.IsSigned $.AllowEmptyPr (not .Repository.IsArchived) .PageIsComparePull}}
 			<div class="ui segment">{{ctx.Locale.Tr "repo.pulls.nothing_to_compare_and_allow_empty_pr"}}</div>
-			<div class="ui info message show-form-container {{if .Flash}}tw-hidden{{end}}">
-				<button class="ui button primary show-form">{{ctx.Locale.Tr "repo.pulls.new"}}</button>
-			</div>
-			<div class="pullrequest-form {{if not .Flash}}tw-hidden{{end}}">
```
The `flash` is the red error message in case an error occurred.

silverwind replied:
Ok, so if there was an flash (presumably only errors), it would also directly open to the form, so this should be compatible.

delvh reviewed (approved):
Finally.
I was waiting for this PR for too long

silverwind commented:
I'm willing to backport this nice enhancement to 1.22.

lunny commented:
It depends on whether the repository is archived/mirror and whether the compared are two branches.

silverwind commented:
> It depends on whether the repository is archived/mirror and whether the compared are two branches.

The `if` condition on the form already includes the two conditions that were previously on the button, so should be fine.

silverwind commented:
> Many `show-form` classes for `<a class="show-form">` should be also removed IMO. They were no-op from beginning.

Removed them and did some minor styling tweaks on that box:

<img width="1333" alt="Screenshot 2024-04-13 at 22 06 57" src="https://github.com/go-gitea/gitea/assets/115237/995ea351-51a5-4444-b360-911aeb7e1a5e">

silverwind commented:
Ready again.

lunny commented:
But sometimes, you just want to do a compare but not create a pull request?

silverwind commented:
> But sometimes, you just want to do a compare but not create a pull request?

You always have the diff to compare at bottom. I see GitHub UI features the button on compare page as well, but I think under some circumstances the button is hidden and it opens directly to the form:

https://github.com/go-gitea/gitea/compare/main...release/v1.22

I think we don't need such complications, we can make the form sufficiently compact and then the page will work for both purposes.

lunny commented:
> > But sometimes, you just want to do a compare but not create a pull request?
> 
> You always have the diff to compare at bottom. I see GitHub UI features the button on compare page as well, but I think under some circumstances the button is hidden and it opens directly to the form:
> 
> [main...release/v1.22](https://github.com/go-gitea/gitea/compare/main...release/v1.22)
> 
> I think we don't need such complications, we can make the form sufficiently compact and then the page will work for both purposes.

But that will bothering. The previous design, there is only 1 line as a hint, but know whatever you want to create or not create a pull request, you will always see the creating pull request form.

silverwind commented:
Maybe we just need a way to skip the button. For example I can open `/user/repo/pull/new/branch` on github and if there is no PR yet, it will open directly to the form. On gitea, I think it links to the compare page and always required clicking the button.

silverwind commented:
Checked how github does it:

```
https://github.com/silverwind/gitea/pull/new/rn
```
Redirects to

```
https://github.com/go-gitea/gitea/compare/main...silverwind:gitea:rn?expand=1
```

So we just need to implement the `expand` option and use it in places where it makes sense to open directly to the form.

wxiaoguang commented:
-> Auto expand "New PR" form #33971

wxiaoguang commented:
Thank you for the initial work, the GitHub approach (`/pull/new/branch-name` redirects `?expand=1`) is implemented in #33971
